src/java/org/apache/cassandra/net/InboundConnectionInitiator.java

```diff
@@ -198,6 +204,54 @@ public static ChannelFuture bind(InboundConnectionSettings settings, ChannelGrou
         return bind(new Initializer(settings, channelGroup, pipelineInjector));
     }
 
+    /**
+     * Handler to perform authentication for internode inbound connections.
+     * This handler is called even before messaging handshake starts.
+     */
+    private static class ClientAuthenticationHandler extends ByteToMessageDecoder
+    {
+        private final IInternodeAuthenticator authenticator;
+
+        public ClientAuthenticationHandler(IInternodeAuthenticator authenticator)
+        {
+            this.authenticator = authenticator;
+        }
+
+        @Override
+        protected void decode(ChannelHandlerContext channelHandlerContext, ByteBuf byteBuf, List<Object> list) throws Exception
+        {
+            // Extract certificates from SSL handler(handler with name "ssl").
+            final Certificate[] certificates = CertificateUtils.certificates(channelHandlerContext.channel());
+            if (!authenticate(channelHandlerContext.channel().remoteAddress(), certificates))
+            {
+                logger.error("Unable to authenticate peer {} for internode authentication", channelHandlerContext.channel());
+                channelHandlerContext.close();
```

Comment:
We should make sure any buffered data is consumed here, so handlers further down the pipeline don't have to deal with unauthenticated payloads after this handler is removed.

A bit like `failHandshake()`, but you could use `channel.pipeline().names()` to remove all pipelines after this one (and leave the exception logging intact).